src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,422 @@
+									   " WHERE worker_hash(%s)>=%ld AND worker_hash(%s)<=%ld) s",
+						 distributionColumnName, tableName,
+						 shardSize,
+						 distributionColumnName, shardMinValue,
+						 distributionColumnName, shardMaxValue
+						 );
+		ereport(DEBUG4, errmsg("%s", AvgHashQuery->data));
+
+		SPI_connect();
+		SPI_exec(AvgHashQuery->data, 0);
+
+		SPITupleTable *tupletable = SPI_tuptable;
+		HeapTuple tuple = tupletable->vals[0];
+		bool isnull;
+
+		Datum average = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+		int64 isResultNull = 1;
+		if (!isnull)
+		{
+			isResultNull = 0;
+		}
+		SPI_freetuptable(tupletable);
+		SPI_finish();
+
+		if (isResultNull == 0)
+		{
+			ereport(DEBUG4, errmsg("%ld", average));
+			MemoryContextSwitchTo(originalContext);
+			splitPoints = lappend_int(splitPoints, DatumGetInt32(average));
+		}
+	}
+	SPI_finish();
+
+	/*Converting the list into datum for further conversion into Arraytype*/
+	Datum *elements = (Datum *) palloc(sizeof(Datum) * list_length(splitPoints));
+	int32 splitPoint, index = 0;
+	foreach_int(splitPoint, splitPoints)
+	{
+		elements[index++] = Int32GetDatum(splitPoint);
+	}
+	ArrayType *resultArray = construct_array(elements, list_length(splitPoints), INT4OID,
+											 sizeof(int32), true, TYPALIGN_INT);
+	if (list_length(splitPoints) == 0)
+	{
+		PG_RETURN_NULL();
+	}
+	PG_RETURN_ARRAYTYPE_P(resultArray);
+}
+
+
+/*
+ * This function calculates the split points of the shard to
+ * split and then executes the background job for the shard split.
+ */
+bool
+ScheduleShardSplit(ShardInfo shardinfo, char *shardSplitMode, int64 jobId)
+{
+	SPI_connect();
+	StringInfo findSplitPointsQuery = makeStringInfo();
+	appendStringInfo(findSplitPointsQuery,
+					 "SELECT citus_find_shard_split_points(%ld , %ld , %ld)",
+					 shardinfo->shardId,
+					 shardinfo->shardSize,
+					 shardinfo->shardGroupSize);
+	SPI_exec(findSplitPointsQuery->data, 0);
+
+	SPITupleTable *tupletable = SPI_tuptable;
+	HeapTuple tuple = tupletable->vals[0];
+	bool isnull;
+	Datum splitPointArrayDatum = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+
+	if (!isnull)
+	{
+		ereport(DEBUG4, errmsg("%s", GetShardSplitQuery(shardinfo, splitPointArrayDatum,
+														shardSplitMode)->data));
+		ExecuteSplitBackgroundJob(jobId, shardinfo, splitPointArrayDatum, shardSplitMode);
+		SPI_finish();
+		return true;
+	}
+	else
+	{
+		ereport(LOG, errmsg("No Splitpoints for shard split"));
+		SPI_finish();
+		return false;
+	}
+}
+
+
+/*
+ * citus_auto_shard_split_start finds shards whose colocation group has total size greater than threshold
+ * and from that extracts the shard with the maximum size . It isolates the shards if the shard has common
+ * tenant values otherwise it 2 way splits it on the basis of average hash value.
+ *
+ */
+Datum
+citus_auto_shard_split_start(PG_FUNCTION_ARGS)
+{
+	ErrorOnConcurrentOperation();
+	StringInfo query = makeStringInfo();
+
```

Comment:
do we want to do an ANALYZE here? That way, we can be sure pg_stats is updated to reflect large tenants.